Autopilot/Src/IMU.cpp

```diff
@@ -0,0 +1,205 @@
+      write_data(writeData);
+
+      //Verify chip is working properly
+      *buffer = REG_WHO_AM_I | 0x80;
+      uint8_t dataFromSensor[2] = {0x00, 0x00}; // dataFromSensor[0] -> dummy byte (contains SPI address); dataFromSensor[1] -> data byte
+      spi_port->set_slave(0);
+      sensorSuccess = spi_port->exchange_data(buffer, dataFromSensor, 2);
+      spi_port->set_slave(0);
+
+      if(sensorSuccess != STATUS_CODE_OK) {
+         //Debug code here
+      }
+
+      //Place accelerometer and gyroscopoe on standby
+      *setup = 0x3F;
+      *buffer = REG_PWR_MGMT_2 | 0x00;
+      writeData[0] = *buffer; 
+      writeData[1] = *setup; 
+      write_data(writeData); 
+
+      //Disable FIFO
+      *setup = 0x00;
+      *buffer = REG_USER_CTRL | 0x00;
+      writeData[0] = *buffer;
+      writeData[1] = *setup;
+      write_data(writeData); 
+
+      //Disable I2C
+      *setup = 0x40;
+      *buffer = REG_I2C_IF | 0x00;
+      writeData[0] = *buffer;
+      writeData[1] = *setup;
+      write_data(writeData); 
+
+      //Set up gyroscope
+      *setup = 0x10; //Setting to 500 dps max for gyroscope; setting to 4G max for accelerometer
+      *buffer = REG_GYRO_CONFIG | 0x00;
+      writeData[0] = *buffer;
+      writeData[1] = *setup;
+      write_data(writeData); 
+
+      //set up accelerometer
+      *buffer = REG_ACCEL_CONFIG | 0x00; 
+      writeData[0] = *buffer;
+      write_data(writeData); 
+
+      *setup = 0x08;
+      *buffer = REG_ACCEL_CONFIG_2 | 0x00;
+      writeData[0] = *buffer;
+      writeData[1] = *setup;
+      write_data(writeData); 
+  
+      //Turn on gyroscope and accelerometer
+      *setup = 0x00;
+      *buffer = REG_PWR_MGMT_2 | 0x00;
+      writeData[0] = *buffer;
+      writeData[1] = *setup;
+      write_data(writeData); 
+   }
+
+   accelConversionFactor = ACCEL_SENSITIVITY_4G;
+   gyroConversionFactor = GRYO_SENSITIVITY_500;
+   tempConversionFactor = TEMPERATURE_SENSITIVITY;
+}
+
+void ICM20602::write_data(uint8_t *writeData) { 
+   spi_port->set_slave(0); //Sets CS to HIGH
+   spi_port->exchange_data(writeData, 0x00, 1);
+   spi_port->set_slave(0); //Sets CS to LOW
+}
+
+void ICM20602::get_accel_temp_gyro_reading(float *accx, float *accy, float *accz, float *gyrx, float *gyry, float *gyrz, float *temp) {
+   uint8_t raw_data[14];
+   uint8_t dataFromSensor[2] = {0x00, 0x00};
+
+   //Store both High and Low Byte values
+   int16_t shiftedSensorAccX, shiftedSensorAccY, shiftedSensorAccZ, shiftedSensorTemp, shiftedSensorGyroX, shiftedSensorGyroY, shiftedSensorGyroZ; 
+
+   //Uses the sensor registers to get raw data for all sensors
+   *buffer = REG_ACCEL_XOUT_H | 0x80; 
+   
+   for(int i = 0; i < 14; i++) {
+      dataFromSensor[1] = 0x00; 
+      spi_port->set_slave(0); 
+      sensorSuccess = spi_port->exchange_data(buffer, dataFromSensor, 2); 
+      spi_port->set_slave(0); 
+      *buffer = REG_ACCEL_XOUT_H + 1; //Moves onto next register
```

Comment:
Clarification:

If I pass an array with 15 elements into the exchange_data() method for the rx_buffer parameter (the first element will be the dummy byte), will I get the sensor data from the next 14 registers in the following 14 elements of the array? 

In my most recent commit, I structured the code using what I described above. (in case I explained it poorly)

Reply:
THING ABOUT RX IS CORRECT

THING ABOUT TX IS INCORRECT, SEND IN 15 ELEMENT ARRAY!! FIRST ELEMENT IS THE X_OUT_H AND THE REST ARE DUMMY BYTES

Reply:
Also the size_t parameter should be 15